nautobot/extras/models/groups.py

```diff
@@ -56,30 +56,198 @@ class DynamicGroup(OrganizationalModel):
+            # (e.g. model form: `cluster` -> filterset: `cluster_id`) yet are omitted from the
+            # filter form (e.g. filter form has "cluster_id" but not "cluster"). We only want to add
+            # them if-and-only-if they aren't already in `filterform_fields`.
+            if missing_field in dynamic_group_filter_fields:
+                mapped_field = dynamic_group_filter_fields[missing_field]
+                if mapped_field in filterform_fields:
+                    logger.debug(
+                        "Skipping missing form field %s; mapped to %s filter field", missing_field, mapped_field
+                    )
+                    continue
+
+            # If the missing field isn't even in the filterset, move on.
+            try:
+                filterset_field = filterset_fields[missing_field]
+            except KeyError:
+                logger.debug("Skipping %s: doesn't have a filterset field", missing_field)
+                continue
+
+            # Get the missing model form field so we can use it to add to the filterform_fields.
+            modelform_field = modelform_fields[missing_field]
+
+            # Replace the modelform_field with the correct type for the UI. At this time this is
+            # only being done for CharField since in the filterset form this ends up being a
+            # `MultVarCharField` (dynamically generated from from `MultiValueCharFilter`) which is
+            # not correct for char fields.
+            if isinstance(modelform_field, forms.CharField):
+                # Get ready to replace the form field w/ correct widget.
+                new_modelform_field = filterset_field.field
+                new_modelform_field.widget = modelform_field.widget
+
+                # If `required=True` was set on the model field, pop "required" from the widget
+                # attributes. Filter fields should never be required!
+                if modelform_field.required:
+                    new_modelform_field.widget.attrs.pop("required")
+
+                modelform_field = new_modelform_field
+
+            # Carry over the `to_field_name` to the modelform_field.
+            to_field_name = filterset_field.extra.get("to_field_name")
+            if to_field_name is not None:
+                modelform_field.to_field_name = to_field_name
+
+            logger.debug("Added %s (%s) to filter fields", missing_field, modelform_field.__class__.__name__)
+            filterform_fields[missing_field] = modelform_field
+
+        # Reduce down to a final dict of desired fields.
+        return_fields = {}
+        for field_name, filter_field in filterform_fields.items():
+            # Skip excluded fields
+            if field_name.startswith(self.exclude_filter_fields):
+                logger.debug("Skipping excluded filter field: %s", field_name)
+                continue
+
+            return_fields[field_name] = filter_field
+
+        return return_fields
```

Comment:
Would it make sense to cache `return_fields` on the instance to avoid unnecessary re-computation? Or is that unneeded?

Reply:
It probably makes sense, seeing as we've asserted that `content_type` is immutable after create. So once we have an instance that is `present_in_database`, the mapped fields should never change.

Reply:
I turned this into a `@cached_property`